packages/core/src/browser/i18n/language-quick-pick-service.ts

```diff
@@ -0,0 +1,124 @@
+// *****************************************************************************
+
+import { inject, injectable } from 'inversify';
+import { nls } from '../../common/nls';
+import { AsyncLocalizationProvider, LanguageInfo } from '../../common/i18n/localization';
+import { QuickInputService, QuickPickItem, QuickPickSeparator } from '../quick-input';
+import { WindowService } from '../window/window-service';
+
+export interface LanguageQuickPickItem extends QuickPickItem {
+    languageId: string
+    execute?(): Promise<void>
+}
+
+@injectable()
+export class LanguageQuickPickService {
+
+    @inject(QuickInputService) protected readonly quickInputService: QuickInputService;
+    @inject(AsyncLocalizationProvider) protected readonly localizationProvider: AsyncLocalizationProvider;
+    @inject(WindowService) protected readonly windowService: WindowService;
+
+    async pickDisplayLanguage(): Promise<string | undefined> {
+        const quickInput = this.quickInputService.createQuickPick();
+        const installedItems = await this.getInstalledLanguages();
+        const quickInputItems: (QuickPickItem | QuickPickSeparator)[] = [
+            {
+                type: 'separator',
+                label: nls.localize('theia/core/installedLanguages', 'Installed languages')
+            },
+            ...installedItems
+        ];
+        quickInput.items = quickInputItems;
+        quickInput.busy = true;
+        const selected = installedItems.find(item => nls.isSelectedLocale(item.languageId));
+        if (selected) {
+            quickInput.activeItems = [selected];
+        }
+        quickInput.placeholder = nls.localizeByDefault('Configure Display Language');
+        quickInput.show();
+
+        this.getAvailableLanguages().then(availableItems => {
+            if (availableItems.length > 0) {
+                quickInputItems.push({
+                    type: 'separator',
+                    label: nls.localize('theia/core/availableLanguages', 'Available languages')
+                });
+                const installed = new Set(installedItems.map(e => e.languageId));
+                for (const available of availableItems) {
+                    // Exclude already installed languages
+                    if (!installed.has(available.languageId)) {
+                        quickInputItems.push(available);
+                    }
+                }
+                quickInput.items = quickInputItems;
+            }
+        }).finally(() => {
+            quickInput.busy = false;
+        });
+
+        return new Promise(resolve => {
+            quickInput.onDidAccept(async () => {
+                const selectedItem = quickInput.selectedItems[0] as LanguageQuickPickItem;
+                // Some language quick pick items want to install additional languages
+                // We have to await that before returning the selected locale
+                await selectedItem.execute?.();
+                resolve(selectedItem.languageId);
```

Comment:
Best to guard with `if (selectedItem)` or the like. It _should_ always  exist, but in the past, there have been errors related to similar behavior in the color and icon theme quick picks.